File: java-demo/src/main/java/demo4/Ex2.java
package demo4;

import java.util.stream.IntStream;

import akka.actor.ActorSystem;
import akka.stream.ActorMaterializer;
import akka.stream.ClosedShape;
import akka.stream.FlowShape;
import akka.stream.Outlet;
import akka.stream.SinkShape;
import akka.stream.UniformFanInShape;
import akka.stream.UniformFanOutShape;
import akka.stream.javadsl.Broadcast;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.FlowGraph;
import akka.stream.javadsl.Merge;
import akka.stream.javadsl.RunnableGraph;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import scala.runtime.BoxedUnit;

public class Ex2 {
  public static void main(String[] args) throws Exception {
    ActorSystem system = ActorSystem.create("demo2");
    ActorMaterializer materializer = ActorMaterializer.create(system);

    // Let's create a cycle
    //
    // source → merge → printer → bcast → sink.ignore
    //             ↑                ↓
    //             ←←←←←←←←←
    RunnableGraph<BoxedUnit> runnable = RunnableGraph.fromGraph(FlowGraph.create(b -> {
      IntStream numbers = IntStream.iterate(1, x -> x + 1);
      Outlet<Integer> source = b.add(Source.from(() -> numbers.iterator())).outlet();
      UniformFanInShape<Integer, Integer> merge = b.add(Merge.create(2));
      UniformFanOutShape<Integer, Integer> bcast = b.add(Broadcast.create(2));
      FlowShape<Integer, Integer> printer = b.add(Flow.of(Integer.class).map(x -> {
        System.out.println(x);
        return x;
      }));
      SinkShape<Integer> sink = b.add(Sink.ignore());

      b.from(source).viaFanIn(merge).via(printer).viaFanOut(bcast).to(sink);
      b.to(merge).fromFanOut(bcast);
      return ClosedShape.getInstance();
    }));

    runnable.run(materializer);
  }
}

File: scala-demo/src/main/scala/demo4/Ex2.scala
package demo4

import akka.actor.ActorSystem
import akka.stream.ActorMaterializer
import akka.stream._
import akka.stream.scaladsl._
import scala.concurrent.duration._
import akka.stream.scaladsl.FlowGraph.Implicits._

object Ex2 extends App {
  implicit val system = ActorSystem("demo4")
  implicit val materializer = ActorMaterializer()
  import system.dispatcher

  // Let's create a cycle
  //
  // source → merge → printer → bcast → sink.ignore
  //             ↑                ↓
  //             ←←←←←←←←←
  val source = Source(() => Iterator.from(1))
  val g = RunnableGraph.fromGraph(
    FlowGraph.create() { implicit b =>
      val merge = b.add(Merge[Int](2))
      val bcast = b.add(Broadcast[Int](2))
      val printer = Flow[Int].map {e => println(e); e}
      source ~> merge ~> printer ~> bcast ~> Sink.ignore
      merge      <~   bcast

      ClosedShape
    }
  )
  g.run()

  // Show that it deadlocks
  // Then show how to fix it with an unfair PreferredMerge junction
  // Then show how to properly fix it by adding a buffering stage
}